src/util/range.h

```diff
@@ -299,7 +307,8 @@ struct zip_iteratort
     ++first_begin;
     ++second_begin;
     INVARIANT(
-      (first_begin == first_end) == (second_begin == second_end),
+      !same_size ||
+        ((first_begin == first_end) == (second_begin == second_end)),
```

Comment:
:bulb: Add an instance of `cbmc_invariants_should_throwt` to your unit test and then test that the invariant is violated when you attempt to increment the iterator beyond the end for various combinations of

- First collection shorter.
- Second collection shorter.
- Both collections the same length with `is_same` set to `true` and `is_same` set to `false`.

It is getting more difficult to see that the invariants are correct by inspection. Therefore it would be good to unit test them.